rust/rust.cc

```diff
@@ -0,0 +1,323 @@
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "log.h"
+#include "nextpnr.h"
+
+namespace {
+    USING_NEXTPNR_NAMESPACE;
+
+    template<typename T>
+    uint64_t wrap(T thing) {
+        static_assert(sizeof(T) <= 8, "T is too big for FFI");
+        uint64_t b = 0;
+        memcpy(&b, &thing, sizeof(T));
+        return b;
+    }
+
+    BelId unwrap_bel(uint64_t bel) {
+        static_assert(sizeof(BelId) <= 8, "T is too big for FFI");
+        auto b = BelId();
+        memcpy(&b, &bel, sizeof(BelId));
+        return b;
+    }
+
+    PipId unwrap_pip(uint64_t pip) {
+        static_assert(sizeof(PipId) <= 8, "T is too big for FFI");
+        auto p = PipId();
+        memcpy(&p, &pip, sizeof(PipId));
+        return p;
+    }
+
+    WireId unwrap_wire(uint64_t wire) {
+        static_assert(sizeof(WireId) <= 8, "T is too big for FFI");
+        auto w = WireId();
+        memcpy(&w, &wire, sizeof(WireId));
+        return w;
+    }
+}
+
+using DownhillIter = decltype(Context(ArchArgs()).getPipsDownhill(WireId()).begin());
+
+struct DownhillIterWrapper {
+    DownhillIter current;
+    DownhillIter end;
+
+    DownhillIterWrapper(DownhillIter begin, DownhillIter end) : current(begin), end(end) {}
+};
+using UphillIter = decltype(Context(ArchArgs()).getPipsUphill(WireId()).begin());
+
+struct UphillIterWrapper {
+    UphillIter current;
+    UphillIter end;
+
+    UphillIterWrapper(UphillIter begin, UphillIter end) : current(begin), end(end) {}
+};
+
+extern "C" {
+    USING_NEXTPNR_NAMESPACE;
+
+    /*
+        DONE:
+        ctx->bindPip
+        ctx->bindWire
+        ctx->check
+        ctx->debug
+        ctx->estimateDelay
+        ctx->getDelayEpsilon
+        ctx->getPipDstWire
+        ctx->getPipSrcWire
+        ctx->getGridDimX
+        ctx->getGridDimY
+        ctx->id
+        ctx->nameOf
+        ctx->unbindWire
+        ctx->verbose
+
+        UNNECESSARY:
+        ctx->getDelayNS - all FFI calls go through it anyway.
+
+        TODO:
+        ctx->checkPipAvail
+        ctx->checkPipAvailForNet
+        ctx->checkWireAvail
+        ctx->getBelPinType
+        ctx->getBoundPipNet
+        ctx->getBoundWireNet
+        ctx->getNetinfoSinkWire
+        ctx->getNetinfoSinkWires
+        ctx->getNetinfoSourceWire
+        ctx->getPipDelay
+        ctx->getPipLocation
+        ctx->getPipsDownhill
+        ctx->getPipsUphill
+        ctx->getRouteBoundingBox
+        ctx->getWireBelPins
+        ctx->getWireDelay
+        ctx->getWires
+        ctx->getWireType
+        ctx->nameOfPip
+        ctx->nameOfWire
+        ctx->nets
+        ctx->nets.at
+        ctx->nets.size
+        ctx->rng64
+        ctx->setting<bool>
+        ctx->setting<float>
+        ctx->setting<int>
+        ctx->sorted_shuffle
+    */
+
+    void npnr_log_info(const char *const format) { log_info("%s", format); }
```

Comment:
this argument is probably better called something other than `format`

Also, I'm not sure if the `const` after the `*` is actually needed here?